packages/jsii-pacmak/lib/targets/dotnet/dotnetgenerator.ts

```diff
@@ -597,6 +611,174 @@ export class DotNetGenerator extends Generator {
+        param.type,
+        noMangle ? name : `{nameof(${name})}`,
+      );
+
+      if (param.optional) {
+        this.code.closeBlock();
+      }
+    }
+    this.code.unindent('#endif');
+
+    function validate(
+      this: DotNetGenerator,
+      value: string,
+      descr: string,
+      type: spec.TypeReference,
+      parameterName: string,
+    ) {
+      if (spec.isUnionTypeReference(type)) {
+        validateTypeUnion.call(this, value, descr, type, parameterName);
+      } else {
+        const collectionType = type as spec.CollectionTypeReference;
+        if (collectionType.collection.kind === spec.CollectionKind.Array) {
+          validateArray.call(
+            this,
+            value,
+            descr,
+            collectionType.collection.elementtype,
+            parameterName,
+          );
+        } else if (collectionType.collection.kind === spec.CollectionKind.Map) {
+          validateMap.call(
+            this,
+            value,
+            descr,
+            collectionType.collection.elementtype,
+            parameterName,
+          );
+        } else {
+          throw new Error(
+            `Unhandled collection kind: ${spec.describeTypeReference(type)}`,
+          );
+        }
+      }
+    }
+
+    function validateArray(
+      this: DotNetGenerator,
+      value: string,
+      descr: string,
+      elementType: spec.TypeReference,
+      parameterName: string,
+    ) {
+      const varName = `__idx_${descr.replace(/[^a-z0-9_]/gi, '_')}`;
+      this.code.openBlock(
+        `for (int ${varName} = 0 ; ${varName} < ${value}.Length ; ${varName}++)`,
+      );
+      validate.call(
+        this,
+        `${value}[${varName}]`,
+        `${descr}[{${varName}}]`,
+        elementType,
+        parameterName,
+      );
+      this.code.closeBlock();
+    }
+
+    function validateMap(
+      this: DotNetGenerator,
+      value: string,
+      descr: string,
+      elementType: spec.TypeReference,
+      parameterName: string,
+    ) {
+      const varName = `__item_${descr.replace(/[^a-z0-9_]/gi, '_')}`;
+      this.code.openBlock(`foreach (var ${varName} in ${value})`);
+      validate.call(
+        this,
+        `${varName}.Value`,
+        `${descr}[\\"{${varName}.Key}\\"]`,
+        elementType,
+        parameterName,
+      );
+      this.code.closeBlock();
+    }
+
+    function validateTypeUnion(
+      this: DotNetGenerator,
+      value: string,
+      descr: string,
+      type: spec.UnionTypeReference,
+      parameterName: string,
+    ) {
+      this.code.indent('if (');
+      let emitAnd = false;
+      const typeRefs = type.union.types;
+      for (const typeRef of typeRefs) {
+        const prefix = emitAnd ? '&& ' : '';
+        const dotNetType = this.typeresolver.toDotNetType(typeRef);
+        // In the case of double, we test for all standard numeric types of .NET (these implicitly convert).
+        const test =
+          dotNetType === 'double'
```

Comment:
There are two branches here (Elvis - `:?` - operator):
1. If the candidate type is `double`, then we check for everything that .NET will implicitly convert to a `double` for us -- that is, all numeric types
2. Otherwise, we can simply use `is <expected-type>`, as .NET does not reify generics ate compile-time like Java does... So `value is IDictionary<string, SomeType>` actually does what you mean for it to do...

So, if your union is (in TypeScript) `Array<string> | Record<string, string>`, then the .NET test we'd emit would look like (remember the condition checks for when we must throw, so it wants to be `true` when the value is NOT valid):

```c#
!(value is IList<string>) && !(value is IDictionary<string, string>)
```